AssertionError when loading the vertical coordinate of an iris cube.

I am not sure if this is a bug in Biggus or in iris.  If I try to load a cube vertical coordinate directly, like this:

```python
>>> import iris
>>> url = "http://colossus.dl.stevens-tech.edu:8080/thredds/dodsC/latest/Bight_gcmplt.nc"
>>> cube = iris.load_cube(url, 'sea_water_salinity')
>>> cube.coord('sea_surface_height_above_reference_ellipsoid')
<repr(<iris.coords.AuxCoord at 0x7f902daf6750>) failed: AssertionError: >
```

and accessing the points trigger the Exception.

```python
Biggus-0.10.0-py2.7.egg/biggus/__init__.pyc in __init__(self, array1, array2, numpy_op, ma_op)
   2679 
   2680         # TODO: Type-promotion
-> 2681         assert array1.dtype == array2.dtype
   2682         self._array1 = array1
   2683         self._array2 = array2
```

But if I load the `cube.coords()`, then everything works.  [Here](http://nbviewer.ipython.org/urls/gist.githubusercontent.com/ocefpaf/f79b25180950b9d28972/raw/57d04028c5994e8c5c5869c55ed08dcf35300514/coords_bug.ipynb) is a notebook showing this issue.

--- a/biggus/__init__.py
+++ b/biggus/__init__.py
@@ -2742,8 +2742,12 @@ class _Elementwise(ComputedArray):
         # TypeError will be raised if not broadcastable.
         array1, array2 = BroadcastArray.broadcast_arrays(array1, array2)
 
-        # TODO: Type-promotion
-        assert array1.dtype == array2.dtype
+        # Type-promotion - The resultant dtype depends on both the array
+        # dtypes and the operation. Avoid using np.find_common_dtype() here,
+        # as integer division yields a float, whereas standard type coercion
+        # rules with find_common_dtype yields an integer.
+        self._dtype = numpy_op(np.ones(1, dtype=array1.dtype),
+                               np.ones(1, dtype=array2.dtype)).dtype
         self._array1 = array1
         self._array2 = array2
         self._numpy_op = numpy_op
@@ -2751,7 +2755,7 @@ class _Elementwise(ComputedArray):
 
     @property
     def dtype(self):
-        return self._array1.dtype
+        return self._dtype
 
     @property
     def shape(self):

--- a/biggus/tests/unit/test__Elementwise.py
+++ b/biggus/tests/unit/test__Elementwise.py
@@ -1,0 +1,68 @@
+# (C) British Crown Copyright 2015, Met Office
+#
+# This file is part of Biggus.
+#
+# Biggus is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Biggus is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Biggus. If not, see <http://www.gnu.org/licenses/>.
+"""Unit tests for `biggus._Elementwise`."""
+
+import unittest
+
+from itertools import permutations
+
+import numpy as np
+import numpy.ma as ma
+
+from biggus import _Elementwise as Elementwise
+
+
+class Test_dtype(unittest.TestCase):
+    def setUp(self):
+        dtypes = []
+        for dtype in ['int', 'uint', 'float']:
+            dtypes.extend(np.sctypes[dtype])
+        self.pairs = [pair for pair in permutations(dtypes, 2)]
+
+    def common_dtype(self, dtype1, dtype2, op):
+        return op(np.ones(1, dtype=dtype1),
+                  np.ones(1, dtype=dtype2)).dtype
+
+    def _test(self, *ops):
+        for dtype1, dtype2 in self.pairs:
+            a1 = np.ones(1, dtype=dtype1)
+            a2 = np.ones(1, dtype=dtype2)
+            actual = Elementwise(a1, a2, *ops)
+            expected_dtype = self.common_dtype(dtype1, dtype2, ops[0])
+            self.assertEqual(actual.dtype, expected_dtype)
+
+    def test_add(self):
+        self._test(np.add, ma.add)
+
+    def test_sub(self):
+        self._test(np.subtract, ma.subtract)
+
+    def test_multiply(self):
+        self._test(np.multiply, ma.multiply)
+
+    def test_floor_divide(self):
+        self._test(np.floor_divide, ma.floor_divide)
+
+    def test_true_divide(self):
+        self._test(np.true_divide, ma.true_divide)
+
+    def test_power(self):
+        self._test(np.power, ma.power)
+
+
+if __name__ == '__main__':
+    unittest.main()

